Bluetooth: Classic: Enable BIP features

lylezhu2012 opened the pull request:
* Bluetooth: OBEX: Add function bt_obex_add_header_body_or_end_body()

A new function bt_obex_add_header_body_or_end_body() is added to help to add the body/end body header. When the tail room of the buffer is more than the passed body room, and the total length of buffer is not more than the mopl if the body has been added, the header end body will be added. Or, the header body will be added.

* Bluetooth: Classic: OBEX: Expose function `bt_obex_string_is_valid()`

The `bt_obex_string_is_valid()` is used to check whether the sending or receiving UTF-16 string is valid or not.

Expose function "bt_obex_string_is_valid()" to support external checking of UTF-16 strings.

* Bluetooth: classic: OBEX: Expose function `bt_obex_has_header()`

The function `bt_obex_has_header()` is used to check whether the specific header is in the buffer.

It is required by upper layer to check the header from the receiving buffer.

* Bluetooth: classic: Enable BIP features

Add functions to manage BIP transport connect and disconnect both for l2cap and rfcomm.

Add functions to register and unregister primary and secondary OBEX servers.

Add functions to send OBEX connect/disconnect request and response.

Add functions to send OBEX get/put request and response with different type defined by BT_BIP_HDR_TYPE_*.

Add functions to send abort request and response.

* Bluetooth: shell: Add BIP shell commands

Add commands for transport over RFCOMM, including `register-rfcomm`, `connect-rfcomm`, and `disconnect-rfcomm`.

Add commands for transport over L2CAP, including `register-l2cap`, `connect-l2cap`, and `disconnect-l2cap`.

Add command `alloc-buf` and `release-buf` to allocate and release TX buffer.

Add command set `add-header` to add the OBEX headers to allocated TX buffer.

Add command set `client` to send BIP client requests with allocated TX buffer.

Add command set `server` to send BIP responses with allocated TX buffer.

* Bluetooth: Classic: BIP: Check features and functions of initiator

Add function `bt_bip_set_supported_capabilities()` to set the supported caps of responder.

Add function `bt_bip_set_supported_features()` to set the supported features of responder.

Add function `bt_bip_set_supported_functions()` to set the supported functions of responder.

Check the supported features of responder when sending OBEX connect request on initiator side.

Check the supported functions of responder when sending OBEX get/put request on initiator side.

* Bluetooth: Classic: SDP: Support BIP profile

Add define `BT_SDP_SUPPORTED_CAPABILITIES` to define supported capabilities attribute.

Add define `BT_SDP_SUPPORTED_FUNCTIONS` to define supported functions attribute.

* Bluetooth: shell: Support BIP SDP record register and discovery

Add command `reg` to register the BIP responder SDP record.

Add command `set_caps`, `set_features`, and `set_functions` to set the supported capabilities, features, and functions for BIP responder.

Add command `discover` to discovery the BIP responder SDP record. And parse the response data to get the supported features and functions.

Add command `set_feats_funcs` to set the discovered supported features and functions.

* Bluetooth: classic: BIP: Check if rsp func is consistent with req's

The response function should be consistent with the function of request when server responses the request of client.

Check whether the BIP function in the server response function is consistent with the client request.

* Bluetooth: Classic: BIP: Ensure uncompleted request func consistent

For the uncompleted BIP request, the following BIP request function should be same with the previous.

Keep the first request function type, and check it with the following request function type.

lylezhu2012 committed:
Bluetooth: classic: Enable BIP features

Add functions to manage BIP transport connect and disconnect both for
l2cap and rfcomm.

Add functions to register and unregister primary and secondary OBEX
servers.

Add functions to send OBEX connect/disconnect request and response.

Add functions to send OBEX get/put request and response with different
type defined by BT_BIP_HDR_TYPE_*.

Add functions to send abort request and response.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: shell: Add BIP shell commands

Add commands for transport over RFCOMM, including `register-rfcomm`,
`connect-rfcomm`, and `disconnect-rfcomm`.

Add commands for transport over L2CAP, including `register-l2cap`,
`connect-l2cap`, and `disconnect-l2cap`.

Add command `alloc-buf` and `release-buf` to allocate and release TX
buffer.

Add command set `add-header` to add the OBEX headers to allocated TX
buffer.

Add command set `client` to send BIP client requests with allocated
TX buffer.

Add command set `server` to send BIP responses with allocated TX
buffer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: OBEX: Expose function `bt_obex_string_is_valid()`

The `bt_obex_string_is_valid()` is used to check whether the sending
or receiving UTF-16 string is valid or not.

Expose function "bt_obex_string_is_valid()" to support external
checking of UTF-16 strings.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: classic: OBEX: Expose function `bt_obex_has_header()`

The function `bt_obex_has_header()` is used to check whether the
specific header is in the buffer.

It is required by upper layer to check the header from the receiving
buffer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: BIP: Check features and functions of initiator

Add function `bt_bip_set_supported_capabilities()` to set the
supported caps of responder.

Add function `bt_bip_set_supported_features()` to set the supported
features of responder.

Add function `bt_bip_set_supported_functions()` to set the supported
functions of responder.

Check the supported features of responder when sending OBEX connect
request on initiator side.

Check the supported functions of responder when sending OBEX get/put
request on initiator side.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: SDP: Support BIP profile

Add define `BT_SDP_SUPPORTED_CAPABILITIES` to define supported
capabilities attribute.

Add define `BT_SDP_SUPPORTED_FUNCTIONS` to define supported functions
attribute.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: shell: Support BIP SDP record register and discovery

Add command `reg` to register the BIP responder SDP record.

Add command `set_caps`, `set_features`, and `set_functions` to set the
supported capabilities, features, and functions for BIP responder.

Add command `discover` to discovery the BIP responder SDP record. And
parse the response data to get the supported features and functions.

Add command `set_feats_funcs` to set the discovered supported features
and functions.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: classic: BIP: Check if rsp func is consistent with req's

The response function should be consistent with the function of
request when server responses the request of client.

Check whether the BIP function in the server response function is
consistent with the client request.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: OBEX: Add function bt_obex_add_header_body_or_end_body()

A new function bt_obex_add_header_body_or_end_body() is added to help
to add the body/end body header. When the tail room of the buffer is
more than the passed body room, and the total length of buffer is not
more than the mopl if the body has been added, the header end body
will be added. Or, the header body will be added.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

makeshi commented on subsys/bluetooth/host/classic/shell/bip.c:
```diff
@@ -0,0 +1,4249 @@
+error_rsp:
+	err = bt_bip_delete_image_rsp(&bip_app.server, rsp_code, bip_app.tx_buf);
+	if (err != 0) {
+		shell_error(sh, "Fail to send delete_image rsp %d", err);
```
Need net_buf_unref(bip_app.tx_buf);?

makeshi commented on subsys/bluetooth/host/classic/bip.c:
```diff
@@ -1882,9 +1882,24 @@ static int bip_get_or_put_rsp(struct bt_bip_server *server, bool is_get, uint8_t
 		err = bt_obex_get_rsp(&server->_server, rsp_code, buf);
 	} else {
+		if (server->_opcode != BT_OBEX_OPCODE_PUT) {
+			LOG_ERR("Invalid operation %u", server->_opcode);
```
It seems same log message with line 1892.

lylezhu2012 replied:
Fixed.

makeshi commented on subsys/bluetooth/host/classic/bip.c:
```diff
@@ -0,0 +1,1955 @@
+	}
+
+	if (err != 0) {
+		LOG_ERR("Failed to send get/put req %d", err);
```
If failure need retset callback to let upper layer retry?

lylezhu2012 commented on subsys/bluetooth/host/classic/bip.c:
```diff
@@ -0,0 +1,1955 @@
+	}
+
+	memcpy(&client->_uuid, uuid, sizeof(client->_uuid));
+	client->_cb = cb;
```
I think it is not needed. The connection status has been checked at line 1332. If the connection has been broken, the old value of `client->_cb` will been ignored.

lylezhu2012 committed:
Bluetooth: Classic: BIP: Ensure uncompleted request func consistent

For the uncompleted BIP request, the following BIP request function
should be same with the previous.

Keep the first request function type, and check it with the following
request function type.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented on subsys/bluetooth/host/classic/shell/bip.c:
```diff
@@ -0,0 +1,4249 @@
+
+	err = bt_bip_abort_rsp(&bip_app.server, rsp_code, bip_app.tx_buf);
+	if (err != 0) {
+		shell_error(sh, "Fail to send abort rsp %d", err);
```
It is not needed. `bip_app.tx_buf` is allocated by a dedicated shell command `bip alloc-buf`. If the function is failed to send. The buffer can be removed by the dedicataed shell command ``bip release-buf``

lylezhu2012 committed:
Bluetooth: classic: OBEX: Fix action id type issue

The OBEX packet cannot be parsed properly if the action ID header is
included in the packet.

It is caused by the value type issue of the action ID. The action ID
type should be `uint8_t` instead of `uint32_t`.

Fix the issue by change action ID value type from `uint32_t` to
`uint8_t`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>
